certora/spec/KintoID/SanctionsTraits.spec

```diff
@@ -0,0 +1,397 @@
+        env e;
+        calldataarg args;
+        f(e, args);
+    address ownerAfter = ownerOf(tokenID);
+
+    assert ownerAfter != ownerBefore => (ownerBefore == 0 || ownerAfter == 0);
+}
+
+rule mintOnlyNextID(address account, method f) filtered{f -> !viewOrUpgrade(f)} {
+    uint256 tokenID = require_uint256(nextTokenId() + 1);
+
+    requireInvariant TokenBalanceIsZeroOrOne(account);
+    requireInvariant BalanceOfZero();
+
+    uint256 balanceBefore = balanceOf(account);
+    address ownerBefore = ownerOf(tokenID);
+        env e;
+        calldataarg args;
+        f(e, args);
+    uint256 balanceAfter = balanceOf(account);
+    address ownerAfter = ownerOf(tokenID);
+
+    assert balanceAfter > balanceBefore => ownerBefore != ownerAfter;
+    assert balanceAfter > balanceBefore => (
+        f.selector == sig:mintCompanyKyc(IKintoID.SignatureData,uint8[]).selector || 
+        f.selector == sig:mintIndividualKyc(IKintoID.SignatureData,uint8[]).selector
+    );
+}
+
+rule mintToOwnerOnly(bool companyOrIndividual) {
+    env e;
+    uint8[] traits;
+    IKintoID.SignatureData signatureData;
+
+    address account = signatureData.signer;
+    uint256 tokenID = require_uint256(nextTokenId() + 1);
+    address ownerBefore = ownerOf(tokenID);
+    if(companyOrIndividual) {
+        mintCompanyKyc(e, signatureData, traits);
+    }
+    else {
+        mintIndividualKyc(e, signatureData, traits);
+    }
+        
+    address ownerAfter = ownerOf(tokenID);
+
+    assert ownerBefore != ownerAfter;
+    assert ownerAfter == account;
+}
+
+rule onlyKYCCanChangeBalance(address account, method f) filtered{f -> !viewOrUpgrade(f)} {
+    uint256 balanceBefore = balanceOf(account);
+        env e;
+        calldataarg args;
+            f(e,args);
+    uint256 balanceAfter = balanceOf(account);
+
+    assert balanceBefore != balanceAfter => hasRole(KYC_PROVIDER_ROLE(), e.msg.sender);
+}
+
+rule burnByOwnerLiveness(uint256 tokenID) {
+    address owner = ownerOf(tokenID);
+
+    require balanceOf(owner) != 0; /// Can only burn if already minted (thus has balance).
+    require totalSupply() != 0; /// Assuming totalSupply() == sum of balances.
+    uint256 tokenIDEnd;
+    require tokensIndex[tokenIDEnd] == assert_uint256(NumberOfTokens - 1);
+    requireInvariant TokenIndexIsUpToArrayLength(tokenID);
+    requireInvariant TokenIndexIsUpToArrayLength(tokenIDEnd);
+    env e; 
+        require e.msg.sender == owner; /// Owner of token is the msg.sender.
+        require e.msg.value == 0; /// Not a payable function.
+    burn@withrevert(e, tokenID);
+    
+    assert !lastReverted;
+}
+
+rule cannotBurnRightAfterMint(IKintoID.SignatureData signatureData) {
+    bool companyOrIndividual;
+    env e1;
+    env e2;
+    uint8[] traits;
+    if(companyOrIndividual) {
+        mintCompanyKyc(e1, signatureData, traits);
+    }
+    else {
+        mintIndividualKyc(e1, signatureData, traits);
+    }
+    burnKYC@withrevert(e2, signatureData);
```

Comment:
this is not right after, it's a different e, so maybe the rule is different; there has to be another operation in between?

Reply:
Right, it's not right after, but the main reason for a different environment is to allow a different msg.sender to perform it (there could be more than one account with the role).